platform/logger.cpp

```diff
@@ -0,0 +1,265 @@
+ * limitations under the License.
+ */
+
+#if defined(MBED_CONF_RTOS_PRESENT) && !defined(MBED_CONF_ZERO_BUFFER_LOGGING) && !defined(NDEBUG)
+#include <string.h>
+#include "platform/logger.h"
+#include "platform/CircularBuffer.h"
+#include "platform/mbed_wait_api.h"
+#include "hal/ticker_api.h"
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_interface.h"
+#include "rtos/EventFlags.h"
+#include "mbed_events.h"
+
+#define LOG_MAX_BUFFER_SIZE_            (MBED_CONF_LOG_MAX_BUFFER_SIZE/sizeof(LOG_DATA_TYPE_))
+// Circular buffer should always have space for helper functions
+// in same line
+#define LOG_FREE_BUF_FLAG_              0x1u
+
+using namespace rtos;
+using namespace mbed;
+
+static EventQueue *log_equeue;
+static CircularBuffer <LOG_DATA_TYPE_, LOG_MAX_BUFFER_SIZE_> log_buffer;
+static const ticker_data_t *const log_ticker = get_us_ticker_data();
+static uint32_t log_bytes_lost = 0;
+static EventFlags log_event_flag("logging");
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+static void log_print_data(void)
+{
+    LOG_DATA_TYPE_ data = 0;
+    while (!log_buffer.empty()) {
+        log_buffer.pop(data);
+#if DEVICE_STDIO_MESSAGES
+#if defined (MBED_ID_BASED_TRACING)
+        fprintf(stderr, "0x%x ", data);
+#else
+        fputc(data, stderr);
+#endif
+#endif
+    }
+    log_event_flag.set(LOG_FREE_BUF_FLAG_);
+}
+
+// Push data in critical section, this is for ISR mode and could be lossy
+static void log_push_data_lossy(const LOG_DATA_TYPE_ *data, uint32_t count)
+{
+    MBED_ASSERT(true == core_util_in_critical_section());
+    uint8_t bytes_written = 0;
+    while (count--) {
+        if (log_buffer.full()) {
+            log_buffer.push('*');
+            log_buffer.push('\n');
+            log_bytes_lost += count;
+            break;
+        } else {
+            log_buffer.push(data[bytes_written++]);
+        }
+    }
+}
+
+// Push data in thread mode
+static void log_push_data_blocking(const LOG_DATA_TYPE_ *data, uint32_t count, uint32_t buf_limit)
+{
+    MBED_ASSERT(true == core_util_in_critical_section());
+    uint8_t bytes_written = 0;
+
+    // Get the buffer length and when size more then that is available push all
+    // data in one go
+    if (buf_limit > (LOG_MAX_BUFFER_SIZE_ - log_buffer.size())) {
+        core_util_critical_section_exit();
+        log_event_flag.clear(LOG_FREE_BUF_FLAG_);
+        log_equeue->call(log_print_data);
+        uint32_t flag = log_event_flag.wait_any(LOG_FREE_BUF_FLAG_);
+        MBED_ASSERT(true != (flag & osFlagsError));
+        core_util_critical_section_enter();
+    }
+    while (count--) {
+        if (log_buffer.full()) {
+            log_buffer.push('*');
+            log_buffer.push('\n');
+            log_bytes_lost += count;
+            break;
+        } else {
+            log_buffer.push(data[bytes_written++]);
+        }
+    }
+}
+
+void mbed_logging_start(void)
+{
+    log_equeue = mbed_event_queue();
```

Comment:
The tracing should have exactly 0 bytes RAM & ROM overhead if not enabled. At least on previous OS versions, only the high priority queue was created by default, so this will create a thread with 1KB stack. Why not just #ifdef the code out?

Reply:
Entire logger.cpp is under #ifdef https://github.com/ARMmbed/mbed-os/pull/5965/files/6c933a7c6454151e44ec375fc7b63510020bee3b#diff-ad984075eb183dde5829e96aec8bf5d7R17 . In case of NDEBUG there will be zero overhead.

Also, we can use high priority queue if required.

Reply:
The use of the default queue is questionable as-is, because the print routine is unbounded - it could keep running indefinitely if some other thread is feeding data, and events are supposed to be circa 10ms.

Use of the high-priority event queue is right out - you're taking more than 100us.

The general problem here is that the wait for serial "block" is being pushed out from the context doing the trace to this thread - which makes the performance look good at the callsite, but would be generally very deleterious if it ends up in moved to the global event queues.

I'd say it would be acceptable to use the normal global event queue if it was one line per event - let the event requeue itself for the second line. That would be friendly to other users.

Still, it does commit us to creating the global event queue. It's really not as good a solution for normal thread-based trace as just using buffered serial (UARTSerial) with mbed-trace, which gives you an interrupt-based buffer data pump with no thread overhead.

Really, we only need to have a dedicated thread/event queue to support the downgrade from IRQ for IRQ tracing, right? (Or to provide buffering for bad unbuffered consoles) It really should be optional. (As it is in mbed-trace  - you have the choice of output functions, so can choose whether to use this data pump or not).

Reply:
> I'd say it would be acceptable to use the normal global event queue if it was one line per event 

Instead of `while (!empty)` we can change that to 2 line equivalent content or till new line character is received.